# 01. Import libraries

In [1]:
# Import libraries 
import pandas as pd
import numpy as np
import os

# 02. Using the dataframe df_dep

In [2]:
path = r'C:\Users\myra_\Documents\Instacart Basket Analysis 02-2023'

In [3]:
# Import the dataset 'departments.csv'
df_dep = pd.read_csv(os.path.join(path, '02 Data', 'Original data', 'departments.csv'), index_col = False)

In [4]:
df_dep.head()

department_id       1      2       3        4        5              6  \
0    department  frozen  other  bakery  produce  alcohol  international   

           7     8                9  ...            12      13         14  \
0  beverages  pets  dry goods pasta  ...  meat seafood  pantry  breakfast   

             15          16         17      18      19    20       21  
0  canned goods  dairy eggs  household  babies  snacks  deli  missing  

[1 rows x 22 columns]

## a) Transposing the dataframe df_dep & refining it

In [5]:
# Transposing df_dep
df_dep.T

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [6]:
# Creating new dataframe from the transposed version
df_dep_T = df_dep.T

In [7]:
df_dep_T

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [8]:
# Creating an index for the dataframe
df_dep_T.reset_index()

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

In [9]:
# Taking the first row of the dataframe as header
new_header = df_dep_T.iloc[0]

In [10]:
new_header

0    department
Name: department_id, dtype: object

In [11]:
# Copying everything from first row onwards to the new dataframe
df_dep_T_new = df_dep_T[1:]

In [12]:
# Making new header for the dataframe
df_dep_T_new.columns = new_header

In [13]:
df_dep_T_new

department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

## b) Exporting df_dep_T_new as a csv file

In [14]:
# Making a csv file for df_dep_T_new
df_dep_T_new.to_csv(os.path.join(path, '02 Data','Prepared data', 'departments_wrangled.csv'))

## c) Making a data dictionary of df_dep_T_new

In [15]:
# turning the dataframe into into a dictionary
data_dict = df_dep_T_new.to_dict('index')

In [16]:
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

# 03. Using the dataframe df_prods

In [17]:
# Importing the 'products.csv' file into Jupyter as df_prods
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original data', 'products.csv'), index_col = False)

In [38]:
df_prods.shape

(49693, 5)

In [18]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [19]:
# To find out what does '19' in the department_id column represent
print(data_dict.get('19'))

{'department': 'snacks'}


In [20]:
# Creating a subset for the df_prods dataframe that only contains data from the snacks department
df_snacks =  df_prods[df_prods['department_id']==19]

In [21]:
df_snacks.head()

product_id                                 product_name  aisle_id  \
0            1                   Chocolate Sandwich Cookies        61   
15          16                Mint Chocolate Flavored Syrup       103   
24          25      Salted Caramel Lean Protein & Fiber Bar         3   
31          32                Nacho Cheese White Bean Chips       107   
40          41  Organic Sourdough Einkorn Crackers Rosemary        78   

    department_id  prices  
0              19     5.8  
15             19     5.2  
24             19     1.9  
31             19     4.9  
40             19     6.5

In [22]:
# To find out what does '4' in the department_id column represent
print(data_dict.get('4'))

{'department': 'produce'}


The 4 in the 'department_id' column represents 'produce'. 

## a) Creating subset of breakfast item sales

In [23]:
# Creating a subset for the df_prods dataframe that only contains data from the breakfast item sales
df_breakfast =  df_prods[df_prods['department_id']==14]

In [24]:
df_breakfast # the below table shows information regarding breakfast item sales

product_id                                       product_name  \
27             28                                  Wheat Chex Cereal   
33             34                                                NaN   
67             68                            Pancake Mix, Buttermilk   
89             90                                       Smorz Cereal   
210           211   Gluten Free Organic Cereal Coconut Maple Vanilla   
...           ...                                                ...   
49330       49326                            Cereal Variety Fun Pack   
49395       49391            Light and Fluffy Buttermilk Pancake Mix   
49547       49543                          Chocolate Cheerios Cereal   
49637       49633               Shake 'N Pour Buttermilk Pancake Mix   
49667       49663  Ultra Protein Power Crunch Peanut Butter N' Ho...   

       aisle_id  department_id  prices  
27          121             14    10.1  
33          121             14    12.2  
67          130             14    13.7  
89          121             14     3.9  
210         130             14     3.6  
...         ...            ...     ...  
49330       121             14     9.1  
49395       130             14     2.0  
49547       121             14    10.8  
49637       130             14    14.2  
49667        57             14    10.2  

[1116 rows x 5 columns]

## b) Creating subset for alcohol, deli, beverages, and meat/seafood

In [25]:
df_dinner_parties = df_prods.loc[df_prods['department_id'].isin([5,7,12,20])]

In [26]:
df_dinner_parties # The following table is a subset with more than one items from the department:

product_id                                    product_name  aisle_id  \
2               3            Robust Golden Unsweetened Oolong Tea        94   
6               7                  Pure Coconut Water With Orange        98   
9              10  Sparkling Orange Juice & Prickly Pear Beverage       115   
10             11                               Peach Mango Juice        31   
16             17                               Rendered Duck Fat        35   
...           ...                                             ...       ...   
49676       49672                          Cafe Mocha K-Cup Packs        26   
49679       49675             Cinnamon Dolce Keurig Brewed K Cups        26   
49680       49676                          Ultra Red Energy Drink        64   
49686       49682                              California Limeade        98   
49688       49684       Vodka, Triple Distilled, Twist of Vanilla       124   

       department_id  prices  
2                  7     4.5  
6                  7     4.4  
9                  7     8.4  
10                 7     2.8  
16                12    17.1  
...              ...     ...  
49676              7     6.5  
49679              7    14.0  
49680              7    14.5  
49686              7     4.3  
49688              5     5.3  

[7650 rows x 5 columns]

The above dataframe has 7650 rows. 

# 04. Using the dataframe df_ords

## a) Changing the variable type

In [27]:
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Original data', 'orders.csv'), index_col = False)

In [28]:
# Changing the data type of order_id to string
df_ords['order_id'] = df_ords['order_id'].astype('str')

In [29]:
df_ords['order_id'].dtype # the dtype for 'order_id' is 0, which means it is a string

dtype('O')

## b) Chaning the column name

In [30]:
df_ords.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0  2539329        1    prior             1          2                  8   
1  2398795        1    prior             2          3                  7   
2   473747        1    prior             3          3                 12   
3  2254736        1    prior             4          4                  7   
4   431534        1    prior             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [31]:
# Renaming the column name and saving it in new dataframe
df_ords_new = df_ords.rename(columns = {'order_dow' : 'orders_day_of_week'})

In [32]:
df_ords_new.head()

order_id  user_id eval_set  order_number  orders_day_of_week  \
0  2539329        1    prior             1                   2   
1  2398795        1    prior             2                   3   
2   473747        1    prior             3                   3   
3  2254736        1    prior             4                   4   
4   431534        1    prior             5                   4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

## c) Finding the frequency

In [33]:
df_ords_new['order_hour_of_day'].value_counts(dropna = False)

10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: order_hour_of_day, dtype: int64

10 AM is the busiest hour for placing the order.

## d) Subset according to customer with user_id = 1

In [34]:
df_user_id_one = df_ords_new.loc[df_ords_new['user_id'].isin([1])]

In [35]:
df_user_id_one

order_id  user_id eval_set  order_number  orders_day_of_week  \
0   2539329        1    prior             1                   2   
1   2398795        1    prior             2                   3   
2    473747        1    prior             3                   3   
3   2254736        1    prior             4                   4   
4    431534        1    prior             5                   4   
5   3367565        1    prior             6                   2   
6    550135        1    prior             7                   1   
7   3108588        1    prior             8                   1   
8   2295261        1    prior             9                   1   
9   2550362        1    prior            10                   4   
10  1187899        1    train            11                   4   

    order_hour_of_day  days_since_prior_order  
0                   8                     NaN  
1                   7                    15.0  
2                  12                    21.0  
3                   7                    29.0  
4                  15                    28.0  
5                   7                    19.0  
6                   9                    20.0  
7                  14                    14.0  
8                  16                     0.0  
9                   8                    30.0  
10                  8                    14.0

In [36]:
df_user_id_one.describe()

user_id  order_number  orders_day_of_week  order_hour_of_day  \
count     11.0     11.000000           11.000000          11.000000   
mean       1.0      6.000000            2.636364          10.090909   
std        0.0      3.316625            1.286291           3.477198   
min        1.0      1.000000            1.000000           7.000000   
25%        1.0      3.500000            1.500000           7.500000   
50%        1.0      6.000000            3.000000           8.000000   
75%        1.0      8.500000            4.000000          13.000000   
max        1.0     11.000000            4.000000          16.000000   

       days_since_prior_order  
count               10.000000  
mean                19.000000  
std                  9.030811  
min                  0.000000  
25%                 14.250000  
50%                 19.500000  
75%                 26.250000  
max                 30.000000

## e) Exporting df_ords as a csv file

In [37]:
df_ords.to_csv(os.path.join(path, '02 Data','Prepared data', 'orders_wrangled.csv'))